src/cogent3/core/new_alignment.py

```diff
@@ -3011,38 +3042,38 @@ class AlignedSeqsData(AlignedSeqsDataABC):
     def __init__(
         self,
         *,
-        seqs: Optional[dict[str, StrORBytesORArray]],
-        gaps: Optional[dict[str, numpy.ndarray]],
+        gapped_seqs: numpy.ndarray,
+        names: tuple[str],
+        ungapped_seqs: Optional[dict[str, numpy.ndarray]] = None,
+        gaps: Optional[dict[str, numpy.ndarray]] = None,
+        offset: Optional[DictStrInt] = None,
         alphabet: new_alphabet.AlphabetABC,
-        offset: dict[str, int] = None,
         align_len: OptInt = None,
         check: bool = True,
     ):
         self._alphabet = alphabet
-        if check or not align_len:
-            if not seqs or not gaps:
-                raise ValueError("Both seqs and gaps must be provided.")
-            if seqs.keys() != gaps.keys():
-                raise ValueError("Keys in seqs and gaps must be identical.")
-            if offset:
-                assert offset.keys() <= seqs.keys()
-            gapped_seq_lengths = {
-                _gapped_seq_len(v, l) for v, l in zip(seqs.values(), gaps.values())
-            }
-            if len(gapped_seq_lengths) != 1:
-                raise ValueError("All sequence lengths must be the same.")
-        self._seqs = {}
-        for k, v in seqs.items():
-            seq = self._alphabet.to_indices(v)
-            self._seqs[k] = seq
-            seq.flags.writeable = False
-        self._gaps = {}
-        for k, v in gaps.items():
-            self._gaps[k] = v
-            v.flags.writeable = False
-        self._align_len = align_len or gapped_seq_lengths.pop()
+        self._names = tuple(names)
+        self._name_to_index = {name: i for i, name in enumerate(names)}
+        self._gapped = gapped_seqs
+        self._ungapped = ungapped_seqs or {}
+        self._gaps = gaps or {}
+        align_len = align_len or gapped_seqs.shape[1]
+        if align_len:
+            assert align_len == gapped_seqs.shape[1], "mismatch in alignment length"
+
+        self._align_len = align_len
         self._offset = offset or {}
 
+        if check:
```

Comment:
we should add a check that the number of names matches the shape of the gapped array